endaq/calc/fft.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import annotations
+
+import typing
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+import scipy.fft
+
+from endaq.calc import psd, utils
+
+
+__all__ = ["aggregate_fft", "fft", "rfft", "dct", "dst"]
+
+
+def aggregate_fft(df, **kwargs):
+    """
+    Calculate the FFT using `scipy.signal.welch` with a specified frequency spacing.  The data returned is in the same
+    units as the data input.
+
+    :param df: the input data
+    :param bin_width: the desired width of the resulting frequency bins, in Hz;
+        defaults to 1 Hz
+    :param kwargs: other parameters to pass directly to ``scipy.signal.welch``
+    :return: a periodogram
+
+    .. seealso::
+
+        - `SciPy Welch's method <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.welch.html>`_
+          Documentation for the periodogram function wrapped internally.
+    """
+    kwargs['scaling'] = 'unit'
+    kwargs['window'] = 'boxcar'
+    kwargs['noverlap'] = 0
+    return psd.welch(df, **kwargs)
+
+
+def fft(
+        df: pd.DataFrame,
+        output: typing.Literal[None, "magnitude", "angle", "complex"] = None,
+        nfft: Optional[int] = None,
+        norm: typing.Literal[None, "unit", "forward", "ortho", "backward"] = None,
+        optimize: bool = True,
+    ) -> pd.DataFrame:
+    """
+    Perform the FFT of the data in `df`, using Scipy's FFT method from `scipy.fft.fft`.  If the in `df` is all real,
+    then the output will be symmetrical between positive and negative frequencies, and it is instead recommended that
+    you use the `rfft` method.
+
+    :param df: the input data
+    :param output: *Optional*  The type of the output of the FFT. Default is "magnitude".  "magnitude" will return the
+                   absolute value of the FFT, "angle" will return the phase angle in radians, "complex" will return the
+                   complex values of the FFT.
+    :param nfft: *Optional* Length of the transformed axis of the output. If nfft is smaller than the length of the
```

Comment:
The `*Optional*` note (here and elsewhere) seems redundant; all keyword arguments are inherently optional. The type hinting also notes it.